api/http/handler/tags/tag_delete.go

```diff
@@ -15,6 +15,55 @@ func (handler *Handler) tagDelete(w http.ResponseWriter, r *http.Request) *httpe
 	if err != nil {
 		return &httperror.HandlerError{http.StatusBadRequest, "Invalid tag identifier route variable", err}
 	}
+	tagID := portainer.TagID(id)
+
+	endpoints, err := handler.EndpointService.Endpoints()
+	if err != nil {
+		return &httperror.HandlerError{http.StatusInternalServerError, "Unable to retrieve endpoints from the database", err}
+	}
+
+	for _, endpoint := range endpoints {
+		endpointTags := []portainer.TagID{}
```

Comment:
I would introduce a new function to delete an element based on index (search and break instead of looping through all tags all the time).

That function could leverage the snippet described in https://github.com/golang/go/wiki/SliceTricks#delete-without-preserving-order

```golang
a[i] = a[len(a)-1]
a[len(a)-1] = nil
a = a[:len(a)-1]
```

Then I would search for the index of the tag to remove, if present remove and update.

Something like this is a lot more readable I think:

```golang
	for _, endpoint := range endpoints {
		tagIdx := 0
		for idx, tag := range endpoint.TagIDs {
			if tag == tagID {
				tagIdx = idx
				break
			}
		}

		if tagIdx != 0 {
			// tag index found, remove and update
		}
	}

```